controller/orphan_controller.go

```diff
@@ -0,0 +1,371 @@
+
+	for _, orphan := range orphans {
+		oc.enqueueOrphan(orphan)
+	}
+}
+
+func (oc *OrphanController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer oc.queue.ShutDown()
+
+	oc.logger.Infof("Start Longhorn Orphan controller")
+	defer oc.logger.Infof("Shutting down Longhorn Orphan controller")
+
+	if !cache.WaitForNamedCacheSync(oc.name, stopCh, oc.cacheSyncs...) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(oc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (oc *OrphanController) worker() {
+	for oc.processNextWorkItem() {
+	}
+}
+
+func (oc *OrphanController) processNextWorkItem() bool {
+	key, quit := oc.queue.Get()
+	if quit {
+		return false
+	}
+	defer oc.queue.Done(key)
+	err := oc.syncOrphan(key.(string))
+	oc.handleErr(err, key)
+	return true
+}
+
+func (oc *OrphanController) handleErr(err error, key interface{}) {
+	if err == nil {
+		oc.queue.Forget(key)
+		return
+	}
+
+	log := oc.logger.WithField("orphan", key)
+
+	if oc.queue.NumRequeues(key) < maxRetries {
+		log.WithError(err).Warnf("Error syncing Longhorn orphan %v: %v", key, err)
+
+		oc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	log.WithError(err).Warnf("Dropping Longhorn orphan %v out of the queue: %v", key, err)
+	oc.queue.Forget(key)
+}
+
+func (oc *OrphanController) syncOrphan(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync orphan %v", oc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != oc.namespace {
+		return nil
+	}
+	return oc.reconcile(name)
+}
+
+func (oc *OrphanController) reconcile(orphanName string) (err error) {
+	orphan, err := oc.ds.GetOrphan(orphanName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	log := getLoggerForOrphan(oc.logger, orphan)
+
+	if !oc.isResponsibleFor(orphan) {
+		return nil
+	}
+
+	if orphan.Status.OwnerID != oc.controllerID {
+		orphan.Status.OwnerID = oc.controllerID
+		orphan, err = oc.ds.UpdateOrphanStatus(orphan)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+		log.Infof("Orphan Controller %v picked up %v", oc.controllerID, orphan.Name)
+	}
+
+	if !orphan.DeletionTimestamp.IsZero() {
+		defer func() {
+			if err == nil {
+				err = oc.ds.RemoveFinalizerForOrphan(orphan)
+			}
+		}()
+
+		// Make sure if the orphan node and controller ID are same.
+		// If NO, just delete the orphan resource object.
+		if orphan.Spec.NodeID != oc.controllerID {
+			return nil
+		}
+
+		node, err := oc.ds.GetNode(orphan.Spec.NodeID)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				log.Debugf("Only delete the orphan resource object since node %v does not exist", node.Name)
+				return nil
+			}
+			return err
+		}
+
+		if orphan.Spec.Type == longhorn.OrphanTypeReplicaDirectory &&
```

Comment:
I feel we should use `longhorn.OrphanTypeReplica` better than `OrphanTypeReplicaDirectory` because it's much straightforward, but it will be not granular a bit for sure.